timdex_dataset_api/record.py

```diff
@@ -16,40 +19,34 @@ class DatasetRecord:
     """
 
     # primary columns
-    timdex_record_id: str
-    source_record: bytes
-    transformed_record: bytes
-
-    # partition columns
-    source: str | None = None
-    run_date: str | datetime.datetime | None = None
-    run_type: str | None = None
-    run_id: str | None = None
-    action: str | None = None
+    timdex_record_id: str = field()
+    source_record: bytes = field()
+    transformed_record: bytes = field()
+    source: str = field()
+    run_date: datetime = field(converter=strict_date_parse)
+    run_type: str = field()
+    run_id: str = field()
+    action: str = field()
+
+    @property
+    def year(self) -> str:
+        return self.run_date.strftime("%Y")
+
+    @property
+    def month(self) -> str:
+        return self.run_date.strftime("%m")
+
+    @property
+    def day(self) -> str:
+        return self.run_date.strftime("%d")
 
     def to_dict(
         self,
-        *,
-        partition_values: dict[str, str | datetime.datetime] | None = None,
-        validate: bool = True,
     ) -> dict:
-        """Serialize instance as dictionary, setting partition values if passed."""
-        if partition_values:
-            for key, value in partition_values.items():
-                setattr(self, key, value)
-        if validate:
-            self.validate()
-        return asdict(self)
-
-    def validate(self) -> None:
-        """Validate DatasetRecord for writing."""
-        # ensure all partition columns are set
-        missing_partition_values = [
-            field
-            for field in ["source", "run_date", "run_type", "run_id", "action"]
-            if getattr(self, field) is None
-        ]
-        if missing_partition_values:
-            raise InvalidDatasetRecordError(
-                f"Partition values are missing: {', '.join(missing_partition_values)}"
-            )
```

Comment:
I think it was the right move to remove `validate()` at this time.  As mentioned during discussions, when we had more complex partitioning going on, and more optional values for a `DatasetRecord`, it felt more important to ensure we had proper partition data.  

But moving into `year -> month -> day` partitioning, and those are dynamic based on a required and parsed field, I feel a lot of that validation is still in place, just moved.